src/diploma.c

```diff
@@ -0,0 +1,218 @@
+
+void CB2_ShowDiploma(void)
+{
+    SetVBlankCallback(NULL);
+    SetGpuReg(REG_OFFSET_DISPCNT, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG3CNT, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG2CNT, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG1CNT, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG0CNT, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG3HOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG3VOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG2HOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG2VOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG1HOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG1VOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG0HOFS, DISPCNT_MODE_0);
+    SetGpuReg(REG_OFFSET_BG0VOFS, DISPCNT_MODE_0);
+    // why doesn't this one use the dma manager either?
+    DmaFill16(3, 0, VRAM, VRAM_SIZE);
+    DmaFill32(3, 0, OAM, OAM_SIZE);
+    DmaFill16(3, 0, PLTT, PLTT_SIZE);
+    remove_some_task();
+    ResetTasks();
+    ResetSpriteData();
+    ResetPaletteFade();
+    FreeAllSpritePalettes();
+    LoadPalette(gDiplomaPalettes, 0, 64);
+    gDiplomaTilemapPtr = malloc(0x1000);
+    InitDiplomaBg();
+    InitDiplomaWindow();
+    reset_temp_tile_data_buffers();
+    decompress_and_copy_tile_data_to_vram(1, &gDiplomaTiles, 0, 0, 0);
+    while (free_temp_tile_data_buffers_if_possible())
+        ;
+    LZDecompressWram(&gDiplomaTilemap, gDiplomaTilemapPtr);
+    CopyBgTilemapBufferToVram(1);
+    DisplayDiplomaText();
+    BlendPalettes(-1, 16, 0);
+    BeginNormalPaletteFade(-1, 0, 16, 0, 0);
+    EnableInterrupts(1);
+    SetVBlankCallback(VBlankCB);
+    SetMainCallback2(MainCB2);
+    CreateTask(Task_DiplomaFadeIn, 0);
+}
+
+static void MainCB2(void)
+{
+    RunTasks();
+    AnimateSprites();
+    BuildOamBuffer();
+    UpdatePaletteFade();
+}
+
+static void Task_DiplomaFadeIn(u8 taskId)
+{
+    if (!gPaletteFade.active)
+        gTasks[taskId].func = Task_DiplomaWaitForKeyPress;
+}
+
+static void Task_DiplomaWaitForKeyPress(u8 taskId)
+{
+    if (gMain.newKeys & (A_BUTTON | B_BUTTON))
+    {
+        BeginNormalPaletteFade(-1, 0, 0, 16, 0);
+        gTasks[taskId].func = Task_DiplomaFadeOut;
+    }
+}
+
+static void Task_DiplomaFadeOut(u8 taskId)
+{
+    if (!gPaletteFade.active)
+    {
+        Free(gDiplomaTilemapPtr);
+        FreeAllWindowBuffers();
+        DestroyTask(taskId);
+        SetMainCallback2(sub_80861E8);
+    }
+}
+
+static void DisplayDiplomaText(void)
+{
+    if (sub_80C0944())
+    {
+        SetGpuReg(REG_OFFSET_BG1HOFS, DISPCNT_BG0_ON);
+        StringCopy(gStringVar1, gText_DexNational);
+    }
+    else
+    {
+        SetGpuReg(REG_OFFSET_BG1HOFS, DISPCNT_MODE_0);
+        StringCopy(gStringVar1, gText_DexHoenn);
+    }
+    StringExpandPlaceholders(gStringVar4, gText_PokedexDiploma);
+    PrintDiplomaText(gStringVar4, 0, 1);
+    PutWindowTilemap(0);
+    CopyWindowToVram(0, 3);
+}
+
+static const struct BgTemplate gDiplomaBgTemplates[2] =
```

Comment:
Didn't we all agree to put const data at the top before the actual code?

Reply:
I have changed my style to do it before the code. I find it much easier to make it line up when its stated before the first function of use.

Reply:
PS: will update save failed screen to comply.